### PR TITLE
backoff retries

### LINKED ISSUE #5: add backoff retries
### Feature description

gRPC should have backoff retries enabled by default, but we need to create a new stream each time the connection is lost and connected again.

### PR DESCRIPTION
**Pull request by maha-hajja**
### Description

check https://github.com/grpc/grpc-go/issues/6094 for why I'm calling the RPC consistently 
Fixes #5 

### Quick checks:

- [x] There is no other [pull request](https://github.com/conduitio-labs/conduit-connector-grpc-client/pulls) for the same update/change.
- [x] I have written unit tests.
- [x] I have made sure that the PR is of reasonable size and can be easily reviewed.

### CODE REVIEW

File: destination.go [hunk 1]
```diff
@@ -133,6 +152,133 @@ func (d *Destination) Write(ctx context.Context, records []sdk.Record) (int, err
 	return len(records), nil
 }
 
+func (d *Destination) sendRecordWithRetries(ctx context.Context, record *opencdcv1.Record) error {
+	d.streamMutex.Lock()
+	if d.stream == nil {
```
[the remark this answers is not in the document]
Comment (maha-hajja) [reply destination.go:157]: if the reconnection fails, ex: max downtime is reached, then d.stream would be nil because `client.Stream(ctx)` returned an error and a nil stream

File: destination.go [hunk 2]
```diff
@@ -133,6 +152,133 @@ func (d *Destination) Write(ctx context.Context, records []sdk.Record) (int, err
...
+		return nil, err
+	}
+	ack, err := d.stream.Recv()
+	if err == io.EOF || status.Code(err) == codes.Unavailable {
+		d.streamMutex.Unlock()
+		err := d.waitForReadyState(ctx, d.conn.GetState())
+		if err != nil {
+			return nil, err
+		}
+		return d.recvAckWithRetries(ctx)
```
[the remark this answers is not in the document]
Comment (lovromazgon) [reply destination.go:188]: In case of resending records we will need to be careful about the acks that get sent back, otherwise they could break the connector. So let's say we sent over 5 records, get back 2 acks, the connection drops. When the connection is re-established we send the last 3 records again (we didn't get acks for those yet). Now what about the acks that we expect to get back? We can get back 3, 4, 5 or 6 acks, depending on what happened on the server. Maybe it didn't receive the 3 records in the first try, so we will get back just 3 acks for the re-sent records. Maybe it received only 1 record and then the connection dropped, and it will be able to send back the 1 ack for the old record + the 3 acks for the re-sent records.

The fun begins when the connection drops while re-sending records 🙃 This will be hard to get right and stable, I'm fine with opening an issue and tackling deduplication of acks another time.

File: destination.go [hunk 3]
```diff
@@ -133,6 +152,133 @@ func (d *Destination) Write(ctx context.Context, records []sdk.Record) (int, err
...
+			return fmt.Errorf("maxDowntime time reached while waiting for server to reconnect")
+		}
+		currentState = d.conn.GetState()
+		// break loop if state is Ready
+		if currentState == connectivity.Ready {
+			break
+		}
+		select {
+		case err := <-d.errCh:
+			return err
```
[the remark this answers is not in the document]
Comment (maha-hajja) [reply destination.go:212]: since we're now listening to `<-d.t.Dying()` and open `ctx` is attached to the tomb, then we can get rid of the case listening to `<-d.openCtx.Done()` here right? or even just keep one of them

File: destination.go [hunk 4]
```diff
@@ -133,6 +152,133 @@ func (d *Destination) Write(ctx context.Context, records []sdk.Record) (int, err
...
+	ticker := time.NewTicker(d.config.ReconnectDelay)
+	var err error
+	timeoutCtx, cancel := context.WithTimeout(ctx, d.config.MaxDowntime)
+	for {
+		select {
+		case <-ticker.C:
+			d.stream, err = client.Stream(ctx)
+			if err == nil {
+				sdk.Logger(ctx).Info().Msg("connection to the server is restored")
+				cancel()
```
[the remark this answers is not in the document]
Comment (maha-hajja) [reply destination.go:266]: just to defend my case... I had these in a loop before turning it into a `select` 👀

File: destination.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (lovromazgon) [reply destination.go:212]: Right, that should work.

File: destination.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (lovromazgon) [reply destination.go:157]: I see 👍

File: destination.go [hunk 5]
```diff
@@ -133,6 +152,133 @@ func (d *Destination) Write(ctx context.Context, records []sdk.Record) (int, err
...
+			}
+		case <-timeoutCtx.Done():
+			cancel()
+			return fmt.Errorf("maxDowntime is reached while waiting for server to reconnect")
+		case <-ctx.Done():
+			cancel()
+			return fmt.Errorf("connector context is canceled")
+		case <-d.t.Dying():
+			cancel()
+			return fmt.Errorf("tearing down connector")
```
[the remark this answers is not in the document]
Comment (maha-hajja) [reply destination.go:277]: ~this error will be added to the errCh, which will be detected when reading a record or receiving an ack.~ 
after some changes: they all return an error now :D

File: destination.go [hunk 6]
```diff
@@ -97,6 +117,14 @@ func (d *Destination) Open(ctx context.Context) error {
 	if err != nil {
 		return fmt.Errorf("failed to create a bidirectional stream: %w", err)
 	}
+	d.errCh = make(chan error, 1)
+	d.openCtx = ctx
+	// spawn a go routine to monitor the connection status
+	d.t = &tomb.Tomb{}
```
[the remark this answers is not in the document]
Comment (maha-hajja) [reply destination.go:123]: didn't know that this was possible! love it

File: destination.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (maha-hajja) [reply destination.go:191]: > So let's say we sent over 5 records, get back 2 acks, then the connection drops. When the connection is re-established we send the last 3 records again

when the connection is re-established, we'd actually continue waiting for acks, we won't send the records again until we get the acks. (because we're not receiving acks in a go routine for now, we send records then wait for acks)
So maybe the whole behavior needs to be changed? I mean, the current behavior does guarantee that this problem won't happen, but, waiting for acks will be the bottleneck until sending new records

[the remark this answers is not in the document]
Comment (lovromazgon) [reply destination.go:191]: > when the connection is re-established, we'd actually continue waiting for acks, we won't send the records again until we get the acks.

I was describing the behavior I proposed, where the client would resend records for which it didn't receive acks yet. There's currently no guarantee that we'll receive the acks, as the server doesn't have any retrying logic. And even if it did, it could very well happen that the server pipeline was restarted, so it won't send any acks, it will just wait for new records, while the client will wait for acks.

I'm thinking that we might need a handshake before opening the stream if we want to do retries correctly. Meaning that when the connection is established the server sends back the last position (i.e. the last ack) so the client can figure out where it should begin retrying sending records. There are still edge cases though, but I won't go into detail too much, I think it would be better to tackle this separately and think thoroughly about the interaction between the client and server.

---

TL;DR: currently the code successfully re-establishes the connection and re-opens the stream, but it doesn't _necessarily_ resume processing records. We need to figure that part out, even if it's very simple logic that could deliver some records twice, as long as it's predictable and stable.

**Review by lovromazgon — APPROVED** [review]
This wasn't an easy task, but seems like we figured it out 👍 Good work, LGTM!
